### PR TITLE
Support Shell script based image creator (in addition to Daisy)

### PR DESCRIPTION
**Pull request by functicons**

### CODE REVIEW

File: custom-images/README.md [hunk 1]
```diff
@@ -21,7 +21,19 @@ Note that custom image will expire in [60 days](https://cloud.google.com/datapro
 
 ## Generate custom image
 
-To generate a custom image, run the following command:
+To generate a custom image, you can run the following command:
+
+```shell
+python generate_custom_image.py \
+    --image-name <new_custom_image_name>
```
Comment (medb) [inline custom-images/README.md]: add `\` at the end of each line

File: custom-images/generate_custom_image.py [hunk 2]
```diff
@@ -378,6 +355,16 @@ def perform_sanity_checks(args):
     raise Exception("Invalid path to customization script: '{}' is not a file.".format(
         args.customization_script))
 
+  # Check the image doesn't already exist.
+  command = "gcloud compute images describe {} --project={}".format(
+      args.image_name, args.project_id)
+  with open(os.devnull, 'w') as devnull:
+    pipe = subprocess.Popen(
+        [command], stdout=devnull, stderr=devnull, shell=True)
```
Comment (medb) [inline custom-images/generate_custom_image.py]: Is `devnull` a valid value?

File: custom-images/shell_script_generator.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Shell script based image creation workflow generator.
+"""
+
+from datetime import datetime
+
+
+class Generator:
```
Comment (medb) [inline custom-images/shell_script_generator.py]: This looks like maintenance nightmare :)

Is it possible to use [Python Template Engine](https://docs.python.org/3.6/library/string.html#template-strings) instead?

File: custom-images/generate_custom_image.py (hunk 2 above)
Comment (functicons) [reply custom-images/generate_custom_image.py]: It is created at line 361 `with open(os.devnull, 'w') as devnull:`

File: custom-images/shell_script_generator.py (hunk 3 above)
Comment (functicons) [reply custom-images/shell_script_generator.py]: Let's keep submit this version first, and switch to template in another PR.

Comment (functicons) [reply custom-images/shell_script_generator.py]: Done.

File: custom-images/shell_script_generator.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+  if [[ "$disk_created" == "true" && "$vm_created" == "false" ]]; then
+    echo 'Deleting disk.'
+    gcloud compute disks delete {image_name}-install --project={project_id} --zone={zone} -q
+  fi
+
+  echo 'Uploading local logs to GCS bucket.'
+  gsutil -m rsync -r /tmp/{run_id}/logs gs://dagang-custom-images/{run_id}/logs/
+
+  sleep 5
```
Comment (medb) [inline custom-images/shell_script_generator.py]: Why sleep here?

File: custom-images/shell_script_generator.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+        --project={project_id} --zone={zone} -q
+  fi
+
+  if [[ "$disk_created" == "true" && "$vm_created" == "false" ]]; then
+    echo 'Deleting disk.'
+    gcloud compute disks delete {image_name}-install --project={project_id} --zone={zone} -q
+  fi
+
+  echo 'Uploading local logs to GCS bucket.'
+  gsutil -m rsync -r /tmp/{run_id}/logs gs://dagang-custom-images/{run_id}/logs/
```
Comment (medb) [inline custom-images/shell_script_generator.py]: Fix bucket name

File: custom-images/shell_script_generator.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+  sleep 5
+}}
+
+function main() {{
+  echo 'Uploading files to GCS bucket.'
+  gsutil cp \
+      {customization_script} \
+      gs://dagang-custom-images/{run_id}/sources/init_actions.sh
+  gsutil cp run.sh gs://dagang-custom-images/{run_id}/sources/
```
Comment (medb) [inline custom-images/shell_script_generator.py]: Fix bucket name

File: custom-images/test_shell_script_generator.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    gcloud compute instances delete my-image-install         --project=my-project --zone=us-west1-a -q
+  fi
+
+  if [[ "$disk_created" == "true" && "$vm_created" == "false" ]]; then
+    echo 'Deleting disk.'
+    gcloud compute disks delete my-image-install --project=my-project --zone=us-west1-a -q
+  fi
+
+  echo 'Uploading local logs to GCS bucket.'
+  gsutil -m rsync -r /tmp/custom-image-my-image-20190611-160823/logs gs://dagang-custom-images/custom-image-my-image-20190611-160823/logs/
```
Comment (medb) [inline custom-images/test_shell_script_generator.py]: Fix bucket name

File: custom-images/test_shell_script_generator.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+  sleep 5
+}
+
+function main() {
+  echo 'Uploading files to GCS bucket.'
+  gsutil cp       /tmp/my-script.sh       gs://dagang-custom-images/custom-image-my-image-20190611-160823/sources/init_actions.sh
```
Comment (medb) [inline custom-images/test_shell_script_generator.py]: Fix bucket name

File: custom-images/shell_script_generator.py (hunk 4 above)
Comment (functicons) [reply custom-images/shell_script_generator.py]: No special reason, just want the user to notice it is exiting.

File: custom-images/shell_script_generator.py (hunk 5 above)
Comment (functicons) [reply custom-images/shell_script_generator.py]: Done.

File: custom-images/shell_script_generator.py (hunk 6 above)
Comment (functicons) [reply custom-images/shell_script_generator.py]: Done.

File: custom-images/test_shell_script_generator.py (hunk 7 above)
Comment (functicons) [reply custom-images/test_shell_script_generator.py]: Done.

File: custom-images/test_shell_script_generator.py (hunk 8 above)
Comment (functicons) [reply custom-images/test_shell_script_generator.py]: Done.

File: custom-images/shell_script_generator.py (hunk 4 above)
Comment (medb) [reply custom-images/shell_script_generator.py]: How sleeps will help to notice exit? Maybe log some exit message?

Comment (functicons) [reply custom-images/shell_script_generator.py]: Done.

**Pull request merged**